internal/model/decoder.go

```diff
@@ -0,0 +1,42 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/internal/model/serializer"
+	"go.opentelemetry.io/collector/internal/model/translator"
+)
+
+// TracesDecoder decodes bytes into pdata.Traces.
+type TracesDecoder struct {
+	translate translator.TracesDecoder
+	serialize serializer.TracesUnmarshaler
+}
+
+// Decode bytes into pdata.Traces.
+func (t *TracesDecoder) Decode(buf []byte) (pdata.Traces, error) {
+	model, err := t.serialize.UnmarshalTraces(buf)
+	if err != nil {
+		return pdata.NewTraces(), fmt.Errorf("unmarshal failed: %w", err)
```

Comment:
I think you can return invalid pdata here `return pdata.Traces{}, err` since we can document that returned value is invalid.